Remove Snap accounts even if the snap is disabled

danroc opened the pull request:
## **Description**

Currently, when removing a snap, we send a `listAccounts` request to the snap to obtain the list of accounts it owns, then we proceed to remove them one by one. However, if the snap is disabled, we cannot retrieve this list and an error is thrown, blocking the snap removal.

This PR resolves this issue by retrieving the list of accounts from the `SnapKeyring` instead of calling the snap. By doing so, it also addresses another bug: the source of truth about the accounts managed by a snap should be the `SnapKeyring` rather than the snap itself. This is because the snap could potentially spoof addresses owned by other keyrings, leading to their inadvertent removal.

Fixes: https://github.com/MetaMask/accounts-planning/issues/55
Depends on: https://github.com/MetaMask/metamask-extension/pull/20684

## **Manual testing steps**

1. Install Flask, and onboard a SRP or create a new one.
2. Install the [SSK](https://metamask.github.io/snap-simple-keyring/0.2.3/).
3. Create 2 accounts using the SSK.
4. Go to "Settings" > "Snaps" > "MetaMask Simple Keyring Snap".
5. Disabled the snap.
6. Click on "Remove MetaMask SSK".
7. Confirm by clicking on "Remove Snap".
8. The snap should have been removed.
9. Check the console of the extension, you should see the following logs:

    <img width="1337" alt="image" src="https://github.com/MetaMask/metamask-extension/assets/68558152/46506da2-c003-4b6f-b9b4-b407c2303199">

## **Screenshots/Recordings**

https://github.com/MetaMask/metamask-extension/assets/68558152/60690737-f5ed-44a6-9a53-4191ef2042ef

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained:
  - [x] What problem this PR is solving.
  - [x] How this problem was solved.
  - [x] How reviewers can test my changes.
- [x] I’ve indicated what issue this PR is linked to: Fixes #???
- [x] I’ve documented any added code.
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
- [x] I’ve properly set the pull request status:
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [x] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [x] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

gantunesr commented on app/scripts/metamask-controller.js:
```diff
@@ -3703,6 +3704,17 @@ export default class MetamaskController extends EventEmitter {
     return address;
   }
 
+  /**
+   * Get the addresses of the accounts managed by a given Snap.
+   *
+   * @param {string} snapId - Snap ID to get accounts for.
+   * @returns {Promise<string[]>} The addresses of the accounts.
+   */
+  async getAccountsBySnapId(snapId) {
+    const snapKeyring = await this.getSnapKeyring();
+    return await snapKeyring.getAccountsBySnapId(snapId);
+  }
+
```
A suggestion for us to consider moving forward, this file is getting too bloated, we should do a refactor to extract some of our logic to another file

owencraston replied:
can this be added to the builder inside [snap-keyring.ts](https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/lib/snap-keyring/snap-keyring.ts)?

gantunesr reviewed (dismissed):
I didn't run the client but the code LGTM

danroc committed:
fix: remove accounts snap even if disabled

Currently, when removing a snap, we send a listAccounts request to the
snap to obtain the list of accounts it owns, then we proceed to remove
them one by one. However, if the snap is disabled, we cannot retrieve
this list and an error is thrown, blocking the snap removal.

This PR resolves this issue by retrieving the list of accounts from the
SnapKeyring instead of calling the snap. By doing so, it also addresses
another bug: the source of truth about the accounts managed by a snap
should be the SnapKeyring rather than the snap itself. This is because
the snap could potentially spoof addresses owned by other keyrings,
leading to their inadvertent removal.

owencraston reviewed (dismissed):
Code looks ✅ 
Ran through the steps and they all were 🤌 
![Screenshot 2023-09-29 at 10 32 13 AM](https://github.com/MetaMask/metamask-extension/assets/22918444/eaf35b41-9590-41c2-977e-65c109e879e9)

danroc committed:
refactor: move `getAccountsBySnapId()` to `snap-keyring.ts`

Mrtenz commented on app/scripts/lib/snap-keyring/snap-keyring.ts:
```diff
@@ -7,6 +7,19 @@ import type {
+ * @param snapId - Snap ID to get accounts for.
+ * @returns The addresses of the accounts.
+ */
+export const getAccountsBySnapId = async (controller: any, snapId: string) => {
```
Can this be typed?

owencraston replied:
I just pushed a commit to address this with better types.

owencraston committed:
better types for getAccountsBySnapId

owencraston commented:
I added better types to the `getAccountsBySnapId` method and confirmed it is still working as expected.
https://github.com/MetaMask/metamask-extension/assets/22918444/6dabb015-7873-4390-9504-7d09c2f33eff

owencraston commented on app/scripts/lib/snap-keyring/snap-keyring.ts:
```diff
@@ -7,6 +7,24 @@ import type {
+  snapId: string,
+) => {
+  const snapKeyring: SnapKeyring =
+    (await controller.getSnapKeyring()) as SnapKeyring;
```
This type cast is needed since the base controller is untyped and the the return type of `getSnapKeyring` is unknown. This could cause errors but i'd argue that throwing an error if we pass in the wrong controller is probably a good thing.

owencraston replied:
I was wrong, by adding the return type in the jsdoc here

```
  /**
   * Initialize the snap keyring if it is not present.
   *
   * @returns {SnapKeyring}
   */
  async getSnapKeyring() {
    let [snapKeyring] = this.coreKeyringController.getKeyringsByType(
      KeyringType.snap,
    );
    if (!snapKeyring) {
      snapKeyring = await this.coreKeyringController.addNewKeyring(
        KeyringType.snap,
      );
    }
    return snapKeyring;
  }
  ```
  
  

Our TS file is able to infer the return type and avoid the cast.

owencraston committed:
annotate return type to avoid casting
